cloud_topics: add l1 object builder + reader

rockwotj opened the pull request:
- **model: add cloud_topics_l1_metadata record batch type**
- **cloud_topics/objects: add l1 builder + reader**

## Backports Required

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

* none

rockwotj committed:
model: add cloud_topics_l1_metadata record batch type

Add new record batch type for tracking L1 object metadata in cloud topics.
This batch type will be used to store markers and indexes that help with
managing the lifecycle and structure of L1 objects in the cloud topics
storage layer.

The change updates the record_batch_type enum to include the new type,
adds the corresponding string representation for logging/debugging, and
updates the MAX value accordingly.

rockwotj committed:
cloud_topics/objects: add l1 builder + reader

Add comprehensive L1 (Level 1) object implementation for cloud topics
storage layer. L1 objects are the foundational building blocks for
organizing and indexing partition data in cloud storage.

The implementation includes:

- object_index: Metadata structure for tracking partitions, offsets,
  timestamps, and file positions within L1 objects. Supports efficient
  seeking by offset or timestamp through indexed entries created
  approximately every 4MiB.

- object_builder: Constructs L1 objects by writing partition markers,
  data batches, and footer indexes. Validates that offsets are strictly
  increasing within partitions and supports configurable indexing
  frequency.

- object_reader: Reads L1 objects sequentially, parsing partition
  markers, data batches, and footer metadata. Can start reading from
  any indexed position for efficient random access.

L1 objects use a well-defined file format with partition markers
(cloud_topics_l1_metadata batches), followed by raft_data batches,
concluded by a footer containing the complete object index. Each
record batch is prefixed with a size field for efficient parsing.

The implementation supports multiple partitions per object with
non-overlapping offset ranges, enables efficient seeking operations,
and includes comprehensive test coverage.

rockwotj commented on src/v/cloud_topics/objects/level_one_object.cc:
```diff
@@ -0,0 +1,355 @@
+protected:
+    void reset_stream(ss::input_stream<char> input) {
+        _input = std::move(input);
+    }
```
Good robot

rockwotj commented:
Force push: remove unused vestigal method

rockwotj commented:
Force pushes: add a check for `read_footer` min buffer length and reformat

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--486586b8-b264-413e-b57c-d1b492abf7ee-->
<!--TEST RESULTS-->
<!--a055416c-008c-43d7-9654-4152e07daffb-->
<!--TEST RESULTS-->
<!--88fb975d-6342-4e07-afd3-1c38414bad3d-->
<!--TEST RESULTS-->
<!--821ddb1d-0866-4f9c-b72b-c14bba62fff5-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67824">build#67824</a></summary>


| test_class                   | test_method                             | test_arguments                       | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:-----------------------------|:----------------------------------------|:-------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageTimingStressTest | test_cloud_storage_with_partition_moves | {"cleanup_policy": "compact,delete"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67824#0197a408-29e6-48c8-a4ba-0b5525d84b88 | FLAKY         | 18/21    | upstream reliability is '98.96373056994818'. current run reliability is '85.71428571428571'. drift is 13.24944 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67899">build#67899</a></summary>


| test_class                  | test_method                             | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------------|:----------------------------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RecreateTopicMetadataTest   | test_recreated_topic_metadata_are_valid | {"replication_factor": 3} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67899#0197a852-bc86-4955-821b-61d7de9dbb4c | FLAKY         | 20/21    | upstream reliability is '97.9035639412998'. current run reliability is '95.23809523809523'. drift is 2.66547 and the allowed drift is set to 50. The test should PASS  |
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test              | {"cloud_storage_type": 1} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67899#0197a850-a3d3-4ed1-981e-f1a6ed538d9d | FLAKY         | 19/21    | upstream reliability is '96.57657657657658'. current run reliability is '90.47619047619048'. drift is 6.10039 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68027">build#68027</a></summary>


| test_class                  | test_method                              | test_arguments                                                                                | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------------|:-----------------------------------------|:----------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ConsumerOffsetsRecoveryTest | test_consumer_offsets_partition_recovery |                                                                                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68027#0197ae20-cdc7-4070-bd32-5d7184d7e9e2 | FLAKY         | 20/21    | upstream reliability is '98.53300733496333'. current run reliability is '95.23809523809523'. drift is 3.29491 and the allowed drift is set to 50. The test should PASS |
| RackAwarePlacementTest      | test_node_config_update                  |                                                                                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68027#0197ae20-cdc7-4448-bd75-69cdd08db994 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| RackAwarePlacementTest      | test_replica_placement                   | {"num_partitions": 50, "num_topics": 2, "rack_layout_str": "ABCDEF", "replication_factor": 3} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68027#0197ae20-cdca-418c-bade-849c4bfa348b | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS             |
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test               | {"cloud_storage_type": 1}                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68027#0197ae20-cdc8-4f7a-8e12-7b40d09728fe | FLAKY         | 20/21    | upstream reliability is '96.70932358318099'. current run reliability is '95.23809523809523'. drift is 1.47123 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68138">build#68138</a></summary>


| test_class                     | test_method                                         | test_arguments                                                                                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------------|:----------------------------------------------------|:-------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageChunkReadTest      | test_read_when_segment_size_smaller_than_chunk_size |                                                                                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68138#0197b4fe-d906-4a62-9f98-a897aa5d17c4 | FLAKY         | 19/21    | upstream reliability is '98.2051282051282'. current run reliability is '90.47619047619048'. drift is 7.72894 and the allowed drift is set to 50. The test should PASS  |
| CompactionTermRollRecoveryTest | test_compact_term_rolled_recovery                   |                                                                                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68138#0197b4fe-d907-43fa-a9ea-64eb194e14b2 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| NodePoolMigrationTest          | test_migrating_redpanda_nodes_to_new_pool           | {"balancing_mode": "node_add", "cleanup_policy": "compact,delete", "test_mode": "tiered_storage_fast_moves"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68138#0197b502-fbdd-46cc-9fb0-cca4948f1ddb | FLAKY         | 20/21    | upstream reliability is '99.42693409742121'. current run reliability is '95.23809523809523'. drift is 4.18884 and the allowed drift is set to 50. The test should PASS |
| PartitionBalancerTest          | test_rack_awareness                                 |                                                                                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68138#0197b502-fbd8-471c-a28b-8cc6401737c5 | FLAKY         | 20/21    | upstream reliability is '98.97435897435898'. current run reliability is '95.23809523809523'. drift is 3.73626 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

dotnwat commented on src/v/model/record_batch_types.h:
```diff
@@ -60,7 +60,8 @@ enum class record_batch_type : int8_t {
     datalake_translation_state = 38, // maintains state for translation progress
     cluster_link = 39,               // cluster link update batches
-    MAX = cluster_link,
+    cloud_topics_l1_metadata = 40,   // markers/indexes for L1 objects
```
using batches for metadata in the L1 object I guess is convenient, but it's a bit weird that we need to reserve a batch type here even though those batches should never be written into a raft group. i guess it doesn't matter, it's just "well known batch types" but maybe a comment about it being L1 only and also not in raft groups might make sense?

mmaslankaprv replied:
It seems to me that it would be much more clear if we wouldn't use the batch for L1 object. Now it is confusing, the object is only uploaded to the cloud

rockwotj replied:
We only need this if we want to support streaming reads without needing the footer. I started a thread in slack, let's talk this out there for higher bandwidth.

rockwotj replied:
Addressed in https://github.com/redpanda-data/redpanda/pull/26563/commits/d9e68065bd32cdc54568142dc1b791dc2fc59305

dotnwat commented on src/v/cloud_topics/objects/level_one_object.h:
```diff
@@ -0,0 +1,380 @@
+// 1. Partition Marker: A metadata record batch containing:
+//    - Key: "pm" (partition_marker_key)
+//    - Value: Serialized model::ntp identifying the partition
+//    - Type: cloud_topics_l1_metadata
```
what's the motivation for the batch type? i mean, the choice of using a batch as a container i guess is the question.

Lazin replied:
I'm also curious why would we need this? We don't need this in the read path because we have a footer with fine grained index.

Lazin replied:
It's OK to have this marker. My question is more about the use case that we have for it.

rockwotj replied:
Let's consolidate this thread with https://github.com/redpanda-data/redpanda/pull/26563#discussion_r2167845194 because they are in some ways the same.

dotnwat commented on src/v/cloud_topics/objects/level_one_object.h:
```diff
@@ -0,0 +1,380 @@
+    struct partition
+      : serde::envelope<partition, serde::version<0>, serde::version<0>> {
+        // The NTP for this partition.
+        model::ntp ntp;
```
i'm guessing we'll want ntp + revision @andrwng ?

Lazin replied:
yep, that's a biggest divergence between the metastore and L1 at the moment

rockwotj replied:
I will update this in a followup (or if Andrew's PR merges first this one, it doesn't need to block - it's a quick update).

andrwng replied:
I've updated my PR to use topic_id and a newly introduced topic_id_partition. Agree no need to block though, should be a quick drop in

dotnwat replied:
i'd think that it still makes sense to keep the ntpr inside the object just to make it easier to look at things with debug tools even if the topic id is the primary key we always use.

rockwotj replied:
I think the footer can use a topic id partition (tip). We can record the ntp in the partition markers (or maybe we reverse it?)

dotnwat commented on src/v/cloud_topics/objects/level_one_object.h:
```diff
@@ -0,0 +1,380 @@
+            // this index entry is located.
+            size_t file_position = 0;
+            // The kafka base_offset of the record batch in this partition at
+            // this position.
```
is "this position" referring to index_entry.file_position or partition.file_position? if the former, i'm a bit confused because the file position would be referring to an index entry batch which wouldn't have a kafka offset.

rockwotj replied:
`index_entry.file_position`

I think this is a misunderstanding that this is an "index entry batch". Which isn't a thing, an index entry points to the batch at which the index entry was generated.

rockwotj replied:
Tried to address in https://github.com/redpanda-data/redpanda/pull/26563/commits/039b81583fe7fad5d306e9297c95a244ebc3084b

dotnwat replied:
>Which isn't a thing, an index entry points to the batch at which the index entry was generated.

Is it accurate to say that an index entry describes a contiguous range of batches, and the file position is the byte offset of the start of that range?

dotnwat commented on src/v/cloud_topics/objects/level_one_object.h:
```diff
@@ -0,0 +1,380 @@
+            bool operator==(const index_entry&) const = default;
+        };
+        // Index information for l1 data, this is a snapshot of the state at a
+        // periodic interval within the partition data. For example, we can have
+        // an index entry every ~4MiB within the partition, since we're working
```
can i perform a single GET and retrieve a 4mb chunk of batches along with the indexing information for that chunk? that random access makes sense, but otherwise i'm not seeing the benefit of spreading the indexing information throughout.

EDIT: i think i see now that the index entry every 4MiB here refers to an index entry at 4MiB granularity, rather than how I had originally read this as finer grained indexing would be located every 4MiB.

rockwotj replied:
Yes exactly, let me reword this so it's clear that we're indexing at that granularity, but we don't mix indexing information and data. All indexing info is in the footer.

rockwotj replied:
Tried to address in https://github.com/redpanda-data/redpanda/pull/26563/commits/039b81583fe7fad5d306e9297c95a244ebc3084b

dotnwat commented on src/v/cloud_topics/objects/level_one_object.h:
```diff
@@ -0,0 +1,380 @@
+
+    // All of partitions in the object, along with their metadata.
+    //
+    // If there are multiple partitions written to an object, then they will
+    // appear multiple times within the vector in the order they where written.
```
👍

bharathv replied:
just to be clear.. this could be multiple partitions objects with same ntp+rev, right? Something like..


```
[Partition 1 Marker][Partition 1 Data][Partition 2 Marker][Partition 2 Data]....[Partition 1 Marker][Partition 1 Data]..
```

rockwotj replied:
Yes exactly. There are no requirements for the partitions within a single L1 object to be unique at the moment.

dotnwat commented on src/v/cloud_topics/objects/level_one_object.h:
```diff
@@ -0,0 +1,380 @@
+    // data.
+    constexpr static size_t npos = std::numeric_limits<size_t>::max();
+
+    // Return the file position of the latest record batch that has the offset
```
>Return the file position of the latest record batch

nit: i'm guessing this isn't strictly true since it would require fine grained indexing? instead it is relaxed "file position of _a_ record batch that has the offset at or before.."?

EDIT: i see in the description below that "batch" is pulling double duty--it's both the mental model for the actual partition data batches, as well as being used to describe the contents of the indexing structure too which makes it a little bit confusing with the terminology clash?

rockwotj replied:
Let me cleanup the wording here.

rockwotj replied:
I tried to address in https://github.com/redpanda-data/redpanda/pull/26563/commits/039b81583fe7fad5d306e9297c95a244ebc3084b

dotnwat commented on src/v/cloud_topics/objects/level_one_object.cc:
```diff
@@ -0,0 +1,353 @@
+        if (partition.ntp != ntp) {
+            continue;
+        }
+        if (target < partition.first_offset || target > partition.last_offset) {
+            continue;
+        }
```
> target < partition.first_offset

i wonder if this is going to be a problem. for example, if I have two L1 objects and I'm looking for an offset that lands in a gap between the ranges contained in the L1 objects then higher level code could know that it should start scanning say the second object at the beginning, but would have to have a special case carved out because it couldn't call `file_position_before_kafka_offset` on the target offset and get back the beginning of the data range?

rockwotj replied:
I think there are a couple of ways of wanting this information when there are multiple partitions within an L1 object (the desired semantics aren't clear to me at this moment), although you're right we should probably fallback to to the first partition after `target` if there is no intervals that contain the offset. I will update.

rockwotj replied:
Addressed in https://github.com/redpanda-data/redpanda/pull/26563/commits/21529107f66f46d848ced12cc758548ee698524a

dotnwat commented on src/v/cloud_topics/objects/level_one_object.cc:
```diff
@@ -0,0 +1,353 @@
+    }
+    iobuf_const_parser parser(buf);
+    std::array<char, sizeof(uint32_t)> footer_size_bytes{};
+    size_t remaining = footer_size_bytes.size();
+    for (auto it = buf.rbegin(); it != buf.rend() && remaining > 0; ++it) {
+        const iobuf::fragment& frag = *it;
+        size_t amt = std::min(remaining, frag.size());
+        remaining -= amt;
+        char* write_at = footer_size_bytes.data();
+        std::advance(write_at, remaining);
+        const char* read_from = frag.get();
+        std::advance(read_from, frag.size() - amt);
+        std::copy_n(read_from, amt, write_at);
+    }
```
yikes hah. i guess we need something like `iobuf iobuf::tail(size)`

rockwotj replied:
Yeah let me pull it out as a helper.

rockwotj replied:
https://github.com/redpanda-data/redpanda/pull/26563/commits/e683073edf24a809c0829fce331c771529a6a50b and https://github.com/redpanda-data/redpanda/pull/26563/commits/6cd5687f92aeb56402a058aff704ab144ebe90cb

dotnwat commented on src/v/cloud_topics/objects/level_one_object.cc:
```diff
@@ -0,0 +1,353 @@
+    template<typename T>
+    ss::future<> serde_write_to_stream(T&& t) {
+        iobuf b;
+        auto placeholder = b.reserve(sizeof(uint32_t));
```
I'm not sure I get the point of the size prefix. From a parsing efficiency perpsective, the record batch header is fixed size and has a field for the total size so we can just pluck that out in O(1) if needed (for scanning or for random reads). if we didn't have the prefix we could use log reader parser or the python storage parser, but if we don't need any of that, what's the point of keeping the metadata in record batches?

rockwotj replied:
I missed the batch header has a `size_bytes` field! I'm not sure how :shy_laugh:. You're right tho we don't need the prefix - I remove this and the serde bits infavor of the fixed batch header like the log reader parser.

rockwotj replied:
Addressed in https://github.com/redpanda-data/redpanda/pull/26563/commits/687515b23341bc7d06b9d12ed02aa694441a5b10

dotnwat reviewed (commented):
yeh looks pretty solid to me. just some random notes/questions reading through this

rockwotj committed:
iobuf: add a tail method

To have an optimized way of getting the last N bytes from an iobuf.

rockwotj committed:
iobuf: use default constructor implementation

We no longer use boost intrusive list

rockwotj committed:
ct/l1: use iobuf::tail in footer read

To remove the fiddly logic to read the last 4 bytes of an iobuf

rockwotj committed:
ct/l1: use fixed header as the prefix delimiter

C++23 is so fun. Instead of using serde and a prefix, we instead write
the header as a fixed width and use the record batch size there to read
the batch data.

We use some fun templates to ensure reading and writing process the
right fields in the right order. Note that we can't share the same code
with local storage because local storage does not store the term
information in the batch header (That's in the filename instead, and
patched in after).

rockwotj commented:
Force push:
- Extract an `iobuf::tail` helper and use that in the `read_footer` method.
- Remove the size prefix from each batch, instead serialize the batch header at a fixed width

rockwotj committed:
ct/l1: use a seperate delimiter for object type

Instead of using a special batch type and a batch as a general
container, use a seperate byte prefix local to l1 objects to distinguish
between the different types.

rockwotj committed:
ct/l1: cleanup documentation

Update the documentation from previous review feedback and also address
some ambiguities from batch types and where indexing information is
stored.

rockwotj committed:
ct/l1: fix offset seeks.

If you seek to an offset before any data, we'll just start at the next
offsets right after the target.

bharathv commented on src/v/cloud_topics/objects/level_one_object.h:
```diff
@@ -0,0 +1,392 @@
+//
+// An L1 object consists of multiple partitions written sequentially, each 
+// containing a series of record batches, followed by a footer with index data.
+//
```
nit: do we need some magic bytes in the beginning that capture format/version etc or maybe that is encoded elsewhere?

rockwotj replied:
My assumption is that version and format is captured in the object path, this gives us the ability to not have to read data to evolve the format (ie if we want to remove/change a header vs footer).

bharathv replied:
ya that makes sense actually.

bharathv commented on src/v/cloud_topics/objects/level_one_object.h:
```diff
@@ -0,0 +1,392 @@
+
+// This file defines the metadata/indexing structure for level one objects. It's
+// written as a footer in the object in cloud storage.
+struct object_index
```
naming nit: Given there are two indexes, one for partitions themselves and one for data within the partition, object_index felt a bit ambiguous when reading.. I was thinking they should be something along the lines of partition_index and partition_data_index ...

rockwotj replied:
I ended up just calling this `footer`. I think it's clearer and more straightforward.

rockwotj replied:
Addressed in https://github.com/redpanda-data/redpanda/pull/26563/commits/7962c0738615826bf1cea013915362551bdd38a5

bharathv commented on src/v/cloud_topics/objects/level_one_object.h:
````diff
@@ -0,0 +1,392 @@
+    // result = l1::object_index::read_footer(std::move(missing));
+    // return std::get<l1::object_index>(result);
+    // ```
+    static std::variant<object_index, size_t> read_footer(iobuf);
````
I'm wondering if footer should have its own serde type (to make it easy to extend in the future). Currently footer == object_index

rockwotj replied:
I just renamed this to footer, I don't see it being worthwhile to add another layer of struct.

Done in https://github.com/redpanda-data/redpanda/pull/26563/commits/7962c0738615826bf1cea013915362551bdd38a5

bharathv commented on src/v/cloud_topics/objects/level_one_object.h:
```diff
@@ -0,0 +1,392 @@
+    //  - A raft data batch, (meaning the header's type is set to raft_data).
+    //  - The offsets in this batch are > than the previous batch in this
+    //    partition.
+    virtual ss::future<> add_batch(model::record_batch) = 0;
```
we may need the capability to write multiple ntps in parallel? Otherwise the API forces the per ntp buffering requirement on to the callers IIUC.

rockwotj replied:
My thinking here is that we can have a method to concat builders, it's not to hard to update the footer, since that is the only places with pointers into the data. I've not added that here, because I'd like to see the need and how to structure the API before creating that method. It's not clear to me how we want to do the combining (in memory, on disk, from cloud directly, etc)

bharathv commented on src/v/cloud_topics/objects/level_one_object.cc:
```diff
@@ -0,0 +1,442 @@
+              p.bytes_left(),
+              size));
+        }
+        return serde::read<object_index>(p);
```
unclear how big these are in practice and if need futurized versions

rockwotj replied:
I'll use futurized versions to be safe, I don't think the extra coroutine here will matter from a perf prospective because this is going to mean you're doing cloud IO.

rockwotj replied:
Addressed in https://github.com/redpanda-data/redpanda/pull/26563/commits/9f158002f224c0a2dfbcf78ade6b76f79b62e229

bharathv commented on src/v/cloud_topics/objects/level_one_object.cc:
```diff
@@ -0,0 +1,442 @@
+    auto min_partition_after_target = partitions.end();
+    for (auto it = partitions.begin(); it != partitions.end(); ++it) {
+        const auto& partition = *it;
+        if (partition.ntp != ntp) {
```
was wondering if the partitions should be a map keyed by ntp since the queries are almost always localized to a single partition

rockwotj replied:
Addressed in https://github.com/redpanda-data/redpanda/pull/26563/commits/7962c0738615826bf1cea013915362551bdd38a5

bharathv commented on src/v/cloud_topics/objects/level_one_object.cc:
```diff
@@ -0,0 +1,442 @@
+      : _input(std::move(input)) {}
+
+    ss::future<result> read_next() final {
+        auto dt_buf = co_await _input.read_exactly(1);
```
nit: maybe pull out these magic numbers (1) to constants

rockwotj replied:
Addressed in https://github.com/redpanda-data/redpanda/pull/26563/commits/7c1fb867e4ba3acfee17e6778809ade676dd6c6e

bharathv reviewed (commented):
🔥 mostly nits except for using a dedicated serde type for footer to extend it in the future.

rockwotj committed:
ct/l1: futurize serde reads

andrwng commented on src/v/cloud_topics/objects/tests/level_one_object_test.cc:
```diff
@@ -0,0 +1,510 @@
+    return reader->read_next().get();
+}
+
+kafka::offset operator""_o(unsigned long long o) {
+    return kafka::offset{static_cast<int64_t>(o)};
+}
+
+model::timestamp operator""_t(unsigned long long t) {
+    return model::timestamp{static_cast<int64_t>(t)};
+}
```
:fire:

rockwotj committed:
ct/l1: remove magic numbers from data type parsing

andrwng commented on src/v/cloud_topics/objects/level_one_object.cc:
```diff
@@ -0,0 +1,447 @@
+        case data_type::partition_marker:
+            co_return co_await read_next_serde<model::ntp>();
+        case data_type::footer:
+            co_return co_await read_next_serde<object_index>();
```
Given this will be based on stored data, add a default case that throws?

rockwotj replied:
Good call, done here: https://github.com/redpanda-data/redpanda/commit/77e7605c0b37ebdcb3cc18e3ad85a57f83cd0bd9

andrwng commented on src/v/cloud_topics/objects/level_one_object.h:
```diff
@@ -0,0 +1,392 @@
+            // The kafka base_offset of the kafka batch that is located at
+            // `file_position`.
+            kafka::offset kafka_offset;
+            // The maximum timestamp seen so far (inclusive of the batch that
+            // generated this index entry) of the kafka batches within the
+            // partition.
+            model::timestamp max_timestamp;
```
Great that we're going to enforce monotonicity here with this. Do you know if the semantics here differ from Kafka's when user input timestamps are out of order? I was under the impression the contract was "find the first timestamp >= than `target`, _regardless_ of whether the timestamps were produced in order, but I'm not sure that's correct

dotnwat replied:
i had a similar thought as well. indexing correctness depends on timestamps being correct, but in reality i figure that we can't force that to be the case so when we see bad timestamps we just get bad indexing?

rockwotj replied:
The indexing isn't bad, the semantics are pretty well defined if not just weird in the face of wildly out of order timestamps.

You need to get the first offset where the timestamp is not before. So keeping the max allows you to prune the space with binary search. If you have an early offset that is way in the future, you just always get that offset back from timequeries.

andrwng reviewed (commented):
Looks great! No major complaints

dotnwat commented on src/v/cloud_topics/objects/level_one_object.cc:
```diff
@@ -0,0 +1,447 @@
+        typename T::type value;
+        std::memcpy(&value, ptr, sizeof(value));
+        return T(value);
+    } else {
+        T value;
+        std::memcpy(&value, ptr, sizeof(value));
+        return value;
+    }
```
don't we need to consider byte ordering? it seems like we'd be better off using serde rather than rolling some new serialization here. easy to change later, for sure so not a blocker, but also surprising to see. is there some motivation for this?

rockwotj replied:
yeah I guess serde is fine here and works as it's fixed width for this stuff, parsing does require an iobuf parser and extra machinery. We will have to use read nested and sort of abuse the framework a bit for the read side

rockwotj replied:
I added byte ordering: https://github.com/redpanda-data/redpanda/pull/26563/commits/77e7605c0b37ebdcb3cc18e3ad85a57f83cd0bd9

I tried to use serde and it's just pretty darn clunky the forced usage of `iobuf` in places like reading a single byte for the data_type or reading 4 bytes for the size prefix in our partition markers. I added a note to this effect. If you feel strongly we should use serde I'm happy to just do it.

dotnwat commented on src/v/cloud_topics/objects/level_one_object.cc:
```diff
@@ -0,0 +1,447 @@
+        if (target > partition.last_offset) {
+            continue;
+        }
+        if (target < partition.first_offset) {
+            if (
+              min_partition_after_target == partitions.end()
+              || partition.first_offset
+                   < min_partition_after_target->first_offset) {
+                min_partition_after_target = it;
+            }
```
do i understand correctly that this is considering that the chunks of a partition within an object may themselves not be in offset order?

rockwotj replied:
yes that is correct. We can change that if we want, but I am also switching to a sorted container

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/68112-->
<!--d41d9b29-a667-4b07-988d-1e772722bf39-->
<!--title-->
## Retry command for [Build#68112](https://buildkite.com/redpanda/redpanda/builds/68112)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/simple_e2e_test.py::SimpleEndToEndTest.test_relaxed_acks@{"write_caching":false}
```
<!--/body-->

dotnwat commented:
merge conflict

rockwotj committed:
ct/l1: index partitions by ntp in footer

A few changes, first of all we rename `object_index` to footer. Then we
use a multimap from ntp to the file positions and index entries for each ntp.

In addition to this, there seems to be some issues in serde with
non-copyable map values, so we don't use serde_fields but instead
specify how to write these values manually.

rockwotj committed:
ct/l1: future proof serialization

Add a note as to why we're not using serde (it's clunky in the read path),
as well as be future proof to forcing `le` value encoding and throwing on unknown data types

rockwotj commented:
Force push: fix merge conflicts

bharathv reviewed (approved):
lgtm pending ci.
